src/blog/wasm-tail-call.md

```diff
@@ -0,0 +1,164 @@
+
+## Using tail calls with Emscripten
+
+Functional languages often depend on tail calls, but it's possible to use them as a C or C++ programmer as well. Emscripten (and Clang, which Emscripten uses) supports the musttail attribute that tells the compiler that a call must be compiled into a tail call. As an example, consider this recursive implementation of a Fibonacci function that calculates the nth Fibonacci number mod 2^32 (because the integers will overflow for large n):
+
+```c
+#include <stdio.h>
+
+unsigned fib_rec(unsigned n, unsigned a, unsigned b) {
+  if (n == 0) {
+    return a;
+  }
+  return fib_rec(n - 1, b, a + b);
+}
+
+unsigned fib(unsigned n) {
+  return fib_rec(n, 0, 1);
+}
+
+int main() {
+  for (unsigned i = 0; i < 10; i++) {
+    printf("fib(%d): %d\n", i, fib(i));
+  }
+
+  printf("fib(1000000): %d\n", fib(1000000));
+}
+```
+
+After compiling with `emcc test.c -o test.js`, running this program in Node.js gives a stack overflow error. We can fix this by adding `__attribute__((__musttail__))` to the return in `fib_rec` and adding `-mtail-call` to the compilation arguments. Now the produced Wasm modules contains the new tail call instructions, so we have to pass `--experimental-wasm-return_call` to Node.js, but the stack no longer overflows.
+
+Here's an example using mutual recursion as well:
+
+```c
+#include <stdio.h>
+#include <stdbool.h>
+
+bool is_odd(unsigned n);
+bool is_even(unsigned n);
+
+bool is_odd(unsigned n) {
+  if (n == 0) {
+    return false;
+  }
+  __attribute__((__musttail__))
+  return is_even(n - 1);
+}
+
+bool is_even(unsigned n) {
+  if (n == 0) {
+    return true;
+  }
+  __attribute__((__musttail__))
+  return is_odd(n - 1);
+}
+
+int main() {
+  printf("is_even(1000000): %d\n", is_even(1000000));
+}
+```
+
+Note that both of these examples are simple enough that if we compile with -O2, the compiler can precompute the answer and avoid exhausting the stack even without tail calls, but this wouldn't be the case with more complex code. In real-world code, the musttail attribute can be helpful for writing high-performance interpreter loops as described in this blog post:
+<https://blog.reverberate.org/2021/04/21/musttail-efficient-interpreters.html>.
+
+Besides the `musttail` attribute, C++ depends on tail calls for one other feature: C++20 coroutines. The relationship between tail calls and C++20 coroutines is covered in extreme depth in this blog post: https://lewissbaker.github.io/2020/05/11/understanding_symmetric_transfer, but to summarize, it is possible to use coroutines in a pattern that would subtly cause stack overflow even though the source code doesn't make it look like there is a problem. To fix this problem, the C++ committee added a requirement that compilers implement "symmetric transfer" to avoid the stack overflow, which in practice means using tail calls under the covers.
```

Comment:
The formatting of the link inside the sentence looks weird. Should we make "this blog post" the link text, and then change it to "a blog post by Lewis Baker"?

Reply:
Done.